src/lavinmq/mqtt/retain_store.cr

```diff
@@ -0,0 +1,173 @@
+        @replicator.register_file(@index_file)
+        @lock = Mutex.new
+        @lock.synchronize do
+          if @index.empty?
+            restore_index(@index, @index_file)
+            write_index
+            @index_file = File.new(@index_file_name, "a+")
+          end
+        end
+      end
+
+      def close
+        @lock.synchronize do
+          write_index
+        end
+      end
+
+      private def restore_index(index : IndexTree, index_file : ::IO)
+        Log.info { "restoring index" }
+        dir = @dir
+        msg_count = 0
+        msg_file_segments = Set(String).new(
+          Dir[Path[dir, "*#{MESSAGE_FILE_SUFFIX}"]].compact_map do |fname|
+            File.basename(fname)
+          end
+        )
+
+        while topic = index_file.gets
+          msg_file_name = make_file_name(topic)
+          unless msg_file_segments.delete(msg_file_name)
+            Log.warn { "msg file for topic #{topic} missing, dropping from index" }
+            next
+          end
+          index.insert(topic, msg_file_name)
+          Log.debug { "restored #{topic}" }
+          msg_count += 1
+        end
+
+        # TODO: Device what's the truth: index file or msgs file. Mybe drop the index file and rebuild
+        # index from msg files?
+        unless msg_file_segments.empty?
+          Log.warn { "unreferenced messages will be deleted: #{msg_file_segments.join(",")}" }
+          msg_file_segments.each do |file_name|
+            File.delete? File.join(dir, file_name)
+          end
+        end
+        Log.info { "restoring index done, msg_count = #{msg_count}" }
+      end
+
+      def retain(topic : String, body_io : ::IO, size : UInt64) : Nil
+        @lock.synchronize do
+          Log.debug { "retain topic=#{topic} body.bytesize=#{size}" }
+          # An empty message with retain flag means clear the topic from retained messages
+          if size.zero?
+            delete_from_index(topic)
+            return
+          end
+
+          unless msg_file_name = @index[topic]?
+            msg_file_name = make_file_name(topic)
+            add_to_index(topic, msg_file_name)
+          end
+
+          tmp_file = File.join(@dir, "#{msg_file_name}.tmp")
+          File.open(tmp_file, "w+") do |f|
+            f.sync = true
+            ::IO.copy(body_io, f)
+          end
+          final_file_path = File.join(@dir, msg_file_name)
+          File.rename(tmp_file, final_file_path)
+          @files.delete(final_file_path)
+          @files[final_file_path] = File.new(final_file_path, "r+")
+          @replicator.replace_file(final_file_path)
+        ensure
+          FileUtils.rm_rf tmp_file unless tmp_file.nil?
+        end
+      end
+
+      private def write_index
+        tmp_file = File.join(@dir, "#{INDEX_FILE_NAME}.next")
+        File.open(tmp_file, "w+") do |f|
+          @index.each do |topic, _filename|
+            f.puts topic
+          end
+        end
+        File.rename tmp_file, @index_file_name
```

Comment:
while not bad with `.next` we've used `.tmp` everywhere else, making it easer to search for in the whole data dir and delete. 

no need for `w+` when you're only writing to a file.

```suggestion
        File.open("#{@index_file_name}.tmp"), "w") do |f|
          @index.each_key do |topic|
            f.puts topic
          end
          f.rename @index_file_name
        end
```